common/llvm.saw

```diff
@@ -1,28 +1,57 @@
 // LLVM utility definitions
 
-// Returns a newly-allocated pointer of the given type that points to
-// the given value.
-let alloc_init ty v = do {
-    p <- crucible_alloc ty;
-    crucible_points_to p v;
-    return p;
+// Remove 'crucible' from common commands
+let points_to = crucible_points_to;
+let precond = crucible_precond;
+let from_cryptol = crucible_term;
+
+let execute = crucible_execute_func;
+let postcond = crucible_postcond;
+let returns = crucible_return;
+let array = crucible_array;
+let llvm_verify module function_name overrides spec =
+  crucible_llvm_verify module function_name overrides true spec z3;
+
+// Declare constants used to denote whether an allocation is
+// 'READONLY' (CONST) or not (WRITE).
+let CONST = {{ False }};
+let WRITE = {{ True }};
+
+// Returns a newly-allocated pointer of the given type.
+let pointer (rw : Term) (type : LLVMType) = if (eval_bool {{ rw == WRITE }})
+                      then crucible_alloc type
+                      else crucible_alloc_readonly type;
+
+//Symbolic value --- for use with the `variable` function.
+let FRESH   = { fresh={{ True }} : Term, t={{ error "NEW Term field was accessed" }} : Term};
+//Concrete value --- for use with the `variable` function.
+let STALE (t : Term) = { fresh={{ False }} : Term, t=t : Term};
+
+let variable_empty = do {
+  return {{ False }};
 };
 
-// Returns a fresh value of the given type along with a newly-allocated
-// pointer that points to that value.
-let ptr_to_fresh n ty = do {
-    x <- crucible_fresh_var n ty;
-    p <- alloc_init ty (crucible_term x);
-    return (x, p);
+// Returns a NULL pointer, the SetupValue, and a Term for either a new
+// symbolic variable or a term passed in.
+// Should be called like:
+//   variable "name" (llvm_int 32) SYM;
+// or
+//   variable "name" (llvm_int 32) (CON {{ 10 : [32] }});
+let variable (type : LLVMType) (name : String) (v : { fresh : Term, t : Term}) = do {
+  cfv <- if (eval_bool v.fresh) then (crucible_fresh_var name type) else variable_empty;
+  let t = if (eval_bool v.fresh) then cfv else v.t;
+  let s = from_cryptol t;
+  return {p=crucible_null, s=s, t=t};
 };
 
-// Returns a fresh value of the given type along with a newly-allocated
-// read-only pointer that points to that value.
-let ptr_to_fresh_readonly n ty = do {
-    x <- crucible_fresh_var n ty;
-    p <- crucible_alloc_readonly ty;
-    crucible_points_to p (crucible_term x);
-    return (x, p);
+// Returns a newly-allocated pointer of the given type that points to
+// a new symbolic variable or a term passed in. Also returns the
+// SetupValue and Term for the variable.
+let alloc (rw : Term) (type : LLVMType) (name : String) (v : { fresh : Term, t : Term}) = do {
+  v' <- variable type name v;
```

Comment:
Similarly, this could be called `alloc_init`, because it _does_ initialize it.

Reply:
I'm trying to write functions that mirror what is done in C. To me, this `alloc` is synonymous with `malloc` (or C++'s `new`), followed by some assignment (`FRESH`or `STALE`).

```
uint32_t *v = malloc(3 * sizeof(uint32_t)); | v <- alloc WRITE (llvm_array 3 uint32_t) "v"
for(i = 0; i < 3; i++) v[i] = i;            |      (STALE {{ [1, 2, 3] : [3][32] }});
double(v);                                  | execute [v.p];
for(i = 0l i < 3; i++) assert(v[i] == i*2); | v' <- variable (llvm_array 3 uint32_t) "v'"
                                            |       (STALE {{ [2, 4, 6] : [3][32] }};
                                            | points_to v.p v'.s;